python/arcticdb/version_store/library.py

```diff
@@ -288,6 +302,195 @@ class ReadInfoRequest(NamedTuple):
     symbol: str
     as_of: Optional[AsOf] = None
 
+    def __repr__(self):
+        res = f"ReadInfoRequest(symbol={self.symbol}"
+        res += f", as_of={self.as_of}" if self.as_of is not None else ""
+        res += ")"
+        return res
+
+
+class LazyDataFrame(QueryBuilder):
+    """
+    Lazy dataframe implementation, allowing chains of queries to be added before the read is actually executed.
+    Returned by `Library.read`, `Library.head`, and `Library.tail` calls when `lazy=True`.
+
+    See Also
+    --------
+    QueryBuilder for supported querying operations.
+
+    Examples
+    --------
+
+    # Specify that we want version 0 of "test" symbol, and to only return the "new_column" column in the output
+    >>> lazy_df = lib.read("test", as_of=0, columns=["new_column"], lazy=True)
+    # Perform a filtering operation
+    >>> lazy_df = lazy_df[lazy_df["col1"].isin(0, 3, 6, 9)]
+    # Create a new column through a projection operation
+    >>> lazy_df["new_col"] = lazy_df["col1"] + lazy_df["col2"]
+    # Actual read and processing happens here
+    >>> df = lazy_df.collect().data
+    """
+    def __init__(
+            self,
+            lib: "Library",
+            read_request: ReadRequest,
+    ):
+        if read_request.query_builder is None:
+            super().__init__()
+        else:
+            self.clauses = read_request.query_builder.clauses
+            self._python_clauses = read_request.query_builder._python_clauses
+            self._optimisation = read_request.query_builder._optimisation
+        self.lib = lib
+        self.read_request = read_request._replace(query_builder=None)
+
+    def to_read_request(self) -> ReadRequest:
```

Comment:
Should we make this private? Why would a user call this?

Reply:
They wouldn't but it's used by `LazyDataFrameCollection`, and IDEs whinge when you use private methods

Reply:
Probably better than having this on the versioned API forever?

Reply:
Yeh fair, have made private